WIP: lexicon learning

xiaohui-zhang opened the pull request:
the main script is  egs/wsj/s5/steps/dict/learn_lexicon.sh

danpovey commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,79 @@
+# decode
+if [ $stage -le 3 ]; then
+  cp -rT data/lang_learned data/lang_learned_rescore
+  cp data/lang_nosp/G.fst data/lang_learned/
```
Unless the words.txt are exactly the same, copying the graphs isn't right.

danpovey replied:
at least check that they are the same with cmp and die if not.

danpovey commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,79 @@
+oov_symbol="<unk>"
+prior_counts="2.5-1-1"
+min_prob=0.3
+var_mass=0.7
```
need comments on all of these variables.  var_mass might not be good name.. self-explanatory names are better.
As separators for strings, commas are better. minus signs could easily be confused with minus signs in numbers e.g. 1.0e-04.

danpovey commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,79 @@
+  utils/prepare_lang.sh --phone-symbol-table data/lang/phones.txt \
+    data/local/dict_learned $oov_symbol data/local/lang_learned data/lang_learned
+  
+  steps/align_si.sh --nj $nj --cmd "$train_cmd" \
```
You shouldn't be using align_si.sh with a SAT model.  You should use align_fmllr.sh if this is the source model.

danpovey commented:
Can you please show us where on the grid this has been run?
@jtrmal, if you have time it would be good if you can look over this.

xiaohui-zhang commented:
here: /export/a11/xzhang/tedlium/s5_r2/exp/tri3_lex_0.4_work  (I added the affix "_0.4" for tuning purposes.) Sure. @jtrmal, I'd like to explain the whole thing to you tmr face to face if you want.

jtrmal commented:
OK, I'll have a look.
y.

On Wed, Oct 5, 2016 at 11:20 PM, Xiaohui Zhang <notifications@github.com>
wrote:

> here: /export/a11/xzhang/tedlium/s5_r2/exp/tri3_lex_0.4_work (I added the
> affix "_0.4" for tuning purposes.) Sure. @jtrmal
> <https://github.com/jtrmal>, I'd like to explain the whole thing to you
> tmr face to face if you want.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-251857928>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX2_MJJMfcvxuDIw-4zz8TbXYtLhjks5qxGkWgaJpZM4JmzJX>
> .
>

jtrmal commented:
Samuel, do these scripts include the yesterday's changes/fixes? I remember there was a script that was crashing.

xiaohui-zhang commented:
@jtrmal Nope. That's Dan's debug_lexicon script which I used to generate alternative pronunciations and that problem only happens for Pashto. I'll suggest changes after I figure out the problem. You can start with my scripts which are more relevant to lexicon learning: local/run_learn_lex.sh, steps/dict/learn_lexicon.sh, steps/dict/select_prons_bayesian.py

jtrmal commented on egs/wsj/s5/steps/dict/select_prons_bayesian.py:
```diff
@@ -0,0 +1,423 @@
+import imp
+import math
+import numpy
+
```
Samuel, I see a bunch of modules you are not actually using here and in the other  python scripts.
Please remove those.  Also, as a general principle, if you importing something outside the default python libs, you should be little bit more helpful with what is wrong.

i.e.
try: 
  import numpy
except ImportError:
  print "Error: you don't seem to have numpy installed, OR the install is not functional"
  raise

btw, please install flake8 and run your script through that program

danpovey commented:
Please avoid using numpy for stuff like this that's not math intensive...
it's not always installed.


On Fri, Oct 7, 2016 at 1:06 PM, jtrmal <notifications@github.com> wrote:

> *@jtrmal* commented on this pull request.
> ------------------------------
>
> In egs/wsj/s5/steps/dict/select_prons_bayesian.py
> <https://github.com/kaldi-asr/kaldi/pull/976#pullrequestreview-3340436>:
>
> > +#!/usr/bin/env python
> +
> +# Copyright 2016  Xiaohui Zhang
> +# Apache 2.0.
> +
> +from __future__ import print_function
> +from collections import defaultdict
> +import os
> +import argparse
> +import sys
> +import warnings
> +import copy
> +import imp
> +import math
> +import numpy
> +
>
> Samuel, I see a bunch of modules you are not actually using here and in
> the other python scripts.
> Please remove those. Also, as a general principle, if you importing
> something outside the default python libs, you should be little bit more
> helpful with what is wrong.
>
> i.e.
> try:
> import numpy
> except ImportError:
> print "Error: you don't seem to have numpy installed, OR the install is
> not functional"
> raise
>
> btw, please install flake8 and run your script through that program
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#pullrequestreview-3340436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu3oM1y-elQ_t-FY_Rk4HmHkFNU44ks5qxnwvgaJpZM4JmzJX>
> .
>

jtrmal commented:
I didn't forget about this. Wanna run the scripts. The thing is the main script is huge, so when I talked with Dan, we were leaning towards splitting it into smaller chunks (or isolating some chunks of the code into separate scripts), but I want to get the feel for it.

BTW, @xiaohui-zhang , you still can address the other things I suggested

xiaohui-zhang commented:
Thanks @jtrmal. Yep I'm addressing issues you found.

On Tue, Oct 11, 2016 at 11:22 AM, jtrmal <notifications@github.com> wrote:

> I didn't forget about this. Wanna run the scripts. The thing is the main
> script is huge, so when I talked with Dan, we were leaning towards
> splitting it into smaller chunks (or isolating some chunks of the code into
> separate scripts), but I want to get the feel for it.
>
> BTW, @xiaohui-zhang <https://github.com/xiaohui-zhang> , you still can
> address the other things I suggested
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-252949790>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANiEEQksj5zRV0U3E2NUT6IDR654Qrsbks5qy6mjgaJpZM4JmzJX>
> .
>



-- 
Xiao-hui

xiaohui-zhang commented:
addressed issues @jtrmal mentioned earlier

jtrmal commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+remove_tags=true
+only_words=true
+icu_transform="Any-Lower"
+cmd=run.pl
```
I think it's OK to remove the whole icu transform mess for the kaldi-wide scripts and perhaps just rely on "tr" or make it user's job?
@danpovey any opinion? I used that to correctly normalize casing of the lexicon (bc for example in Turkish, i is not lowercased I) and was handling utf-8 encoding well. But not sure if/how this is importat for kaldi in general

jtrmal replied:
I just talked with Dan and let's go with ignoring the normalization issues and let's make it user's job.. The same with the apply_g2p.sh
That will simplify the g2p scripts considerably

jtrmal commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+
+[ -f ./path.sh ] && . ./path.sh; # source the path.
+. parse_options.sh || exit 1;
+
```
Might be better to write
```
. utils/parse_options.sh
```

danpovey commented:
@jtrmal and @xiaohui-zhang, can you please remind me what are the next TODOs to get this in shape?

jtrmal commented:
I want to run the scripts on graphemic babel just to see how the scripts
behave...
y.

On Mon, Oct 17, 2016 at 3:45 PM, Daniel Povey <notifications@github.com>
wrote:

> @jtrmal <https://github.com/jtrmal> and @xiaohui-zhang
> <https://github.com/xiaohui-zhang>, can you please remind me what are the
> next TODOs to get this in shape?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-254312435>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX2-W1N2UNCqgKzVwfJ-xOZ0tc0Tzks5q09BXgaJpZM4JmzJX>
> .
>

xiaohui-zhang commented:
it'll be great if someone can run it on tedlium first since English is
understandable and it's easier to figure out potential problems in my
recipe...

On Mon, Oct 17, 2016 at 4:02 PM, jtrmal <notifications@github.com> wrote:

> I want to run the scripts on graphemic babel just to see how the scripts
> behave...
> y.
>
> On Mon, Oct 17, 2016 at 3:45 PM, Daniel Povey <notifications@github.com>
> wrote:
>
> > @jtrmal <https://github.com/jtrmal> and @xiaohui-zhang
> > <https://github.com/xiaohui-zhang>, can you please remind me what are
> the
> > next TODOs to get this in shape?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-254312435>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-
> auth/AKisX2-W1N2UNCqgKzVwfJ-xOZ0tc0Tzks5q09BXgaJpZM4JmzJX>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-254316967>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANiEEZDsBsv7WEKKapEuU7BLLnVQxcOsks5q09RwgaJpZM4JmzJX>
> .
>



-- 
Xiao-hui

jtrmal commented:
Samuel, any progress on implementing the changes?

jtrmal commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,111 @@
+
+
+. utils/parse_options.sh # accept options
+
```
Either use 
```
set -e -o pipefail
```
or make sure all calls ends with
```
|| exit 1
```

jtrmal commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+
+# Begin configuration section.  
+iters=5
+stage=5
```
stage should be set to 0

jtrmal commented:
I played with the scripts and I think the steps/learn_lexicon.sh should really be re-desgined.
Mainly, the train_g2p and apply_g2p should be taken out.
The reason is that in many recipes, you will have g2p already trained (especially given the fact we will shift the responsibility for lexicon normalization for g2p to the user) or you actually do not need the g2p at all -- For graphemic systems,  you can generate pronuncation right away (albeit badly) and in some cases you might decide you don't want to mess with g2p at all and want use only  the prons found in the lattice.

xiaohui-zhang commented:
I agree with @jtrmal that we should definitely take train_g2p out of the recipe. However, I doubt whether we should take apply_g2p out of the recipe. The reason is that, when running phone-decoding (debug_lexicon), we prefer a lexicon which covers most words in acoustic training data, otherwise too many oovs will make phone-decoding very slow. More importantly, in many cases we want to include pronunciation candidates from g2p in the pronunciation selection procedure. But I think we can make apply_g2p optional here, assuming the user-provided reference lexicon already covers most words in acoustic training data, and the user doesn't care about any pron from g2p during pronunciation selection. So I think I can add an option like "use_g2p_for_oov" and change the script accordingly. @danpovey what do you think?

danpovey commented:
[I sent this by email, but github did not post it.]
I think what Yenda was saying was not that you should change what the recipe does, but how it's structured, so that train_g2p and apply_g2p are done by the caller (e.g. in a script in local/) before your script is called.
Is there a reason why that is not feasible, or in your opinion a good idea?

xiaohui-zhang commented:
Thanks. I think it makes sense. In order to take g2p totally out of the steps level script, instead of asking the user provide a "target vocab" which we aim to generate prons for, now I just ask the user to provide the ref lexicon, and make the script just output an lexicon-edit file, containing suggested changes to the ref lexicon. If the user additionally provide a complementary g2p-generated lexicon, which contains words seen in acoustic data but out of the ref lexicon, this script will also output an edit-file for this lexicon (maybe just the learned/improved lexicon, cause we may not expect people the evaluate this part by hand?)

danpovey commented:
[I replied by email, but posting also, because github is not posting my emails for the last few days, for some reason.]

I think the edit file should be with respect to the original lexicon, and the g2p-derived lexicon should be just provided as a set of possible suggestions for prons we might want to add.  So it would behave the same as it does in the existing recipe, just come from the user, not from the script itself.

danpovey commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+[ ! -f $lexicon ] && echo "$0: Training lexicon does not exist." && exit 1
+
+if $only_words ; then
+  cat $lexicon | sed 's/^<.*>.*$//g' | sed 's/^#.*//g' > $wdir/lexicon_onlywords.txt
```
This step is too specific to the BABEL recipe.  A more general way to do this would be to use the same method as in
steps/cleanup/internal/get_non_scored_words.py,
but that script would need a list of silence phones, so you'd have to provide such a list (which would normally be available as it's needed in lexicon dirs).

danpovey commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+
+if $remove_tags ; then
+  cat $lexicon |\
+    sed 's/_[%|"]//g' | sed 's/_[0-9]\+//g' > $wdir/lexicon_notags.txt
```
This, also, is too specific to the BABEL recipe.  I think it would be safer to just leave it out.

danpovey commented on egs/wsj/s5/steps/dict/train_g2p.sh:
```diff
@@ -0,0 +1,94 @@
+  lexicon=$wdir/lexicon_notags.txt
+fi
+
+if [ ! -z $icu_transform ] ; then
```
and as Yenda said, we're leaving icu_transform out.

danpovey commented:
I think what Yenda was saying was not that you should change what the
recipe does, but how it's structured, so that train_g2p and apply_g2p are
done by the caller (e.g. in a script in local/) before your script is
called.
Is there a reason why that is not feasible, or in your opinion a good idea?




On Sat, Oct 22, 2016 at 3:30 PM, Xiaohui Zhang <notifications@github.com>
wrote:

> I agree with @jtrmal <https://github.com/jtrmal> that we should
> definitely take train_g2p out of the recipe. However, I doubt whether we
> should take apply_g2p out of the recipe. The reason is that, when running
> phone-decoding (debug_lexicon), we prefer a lexicon which covers most words
> in acoustic training data, otherwise too many oovs will make phone-decoding
> very slow. More importantly, in many cases we want to include pronunciation
> candidates from g2p in the pronunciation selection procedure. But I think
> we can make apply_g2p optional here, assuming the user-provided reference
> lexicon already covers most words in acoustic training data, and the user
> doesn't care about any pron from g2p during pronunciation selection. So I
> think I can add an option like "use_g2p_for_oov" and change the script
> accordingly. @danpovey <https://github.com/danpovey> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-255548872>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuy2E-0BITi7wVLCJX7Ww6-ZIuV8Iks5q2mRNgaJpZM4JmzJX>
> .
>

danpovey commented:
I think the edit file should be with respect to the original lexicon, and
the g2p-derived lexicon should be just provided as a set of possible
suggestions for prons we might want to add.  So it would behave the same as
it does in the existing recipe, just come from the user, not from the
script itself.



On Sat, Oct 22, 2016 at 4:52 PM, Xiaohui Zhang <notifications@github.com>
wrote:

> Thanks. I think it makes sense. In order to take g2p totally out of the
> steps level script, instead of asking the user provide a "target vocab"
> which we aim to generate prons for, now I just ask the user to provide the
> ref lexicon, and make the script just output an lexicon-edit file,
> containing suggested changes to the ref lexicon. If the user additionally
> provide a complementary g2p-generated lexicon, which contains words seen in
> acoustic data but out of the ref lexicon, this script will also output an
> edit-file for this lexicon (maybe just the learned/improved lexicon, cause
> we may not expect people the evaluate this part by hand?)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-255553320>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu14Jjv1Vz1oEHBAHYNvTkg51LJgFks5q2nePgaJpZM4JmzJX>
> .
>

danpovey commented:
OK, so I guess you're saying that train_g2p and apply_g2p, if needed,
should be done at the local/ level.  And maybe the lexicon-learning script
should take the 'suggested' (g2p-sourced or other-sourced) pronunciations
[maybe with associated priors, like a lexiconp.txt format?] as an
argument.  That should be doable.




On Thu, Oct 20, 2016 at 1:46 PM, jtrmal <notifications@github.com> wrote:

> I played with the scripts and I think the steps/learn_lexicon.sh should
> really be re-desgined.
> Mainly, the train_g2p and apply_g2p should be taken out.
> The reason is that in many recipes, you will have g2p already trained
> (especially given the fact we will shift the responsibility for lexicon
> normalization for g2p to the user) or you actually do not need the g2p at
> all -- For graphemic systems, you can generate pronuncation right away
> (albeit badly) and in some cases you might decide you don't want to mess
> with g2p at all and want use only the prons found in the lattice.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-255177325>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuz11VRSpEpQgFbn1uLoLA9NtsimLks5q16kEgaJpZM4JmzJX>
> .
>

danpovey commented on egs/wsj/s5/steps/cleanup/internal/get_pron_stats.py:
```diff
@@ -0,0 +1,203 @@
+
+# Basically, this function generates an "info" list from a ctm_prons file.
+# Each entry in the list represents the pronounciation candidate(s) of a word.
+# For each non-<eps> word, the entry is a list: [utt_id, word, set(prnounciation_candidates)].
```
typo on this line.
Also, please give an example of the data structure, e.g.
`[ 'foo-bar123', 'hello', set(['h ah l l', 'ha l l ow' ])]` or whatever it is.

danpovey commented on egs/wsj/s5/steps/dict/apply_lexicon_edits.py:
```diff
@@ -0,0 +1,105 @@
+import sys
+
+def GetArgs():
+    parser = argparse.ArgumentParser(description = "Applyccumulate statistics from lattice-alignment"
```
applycumulate?
Also there needs to be a comment at the top (and/or a usage message) that explains what this is  about, where its input comes from, gives examples of the format, etc.

danpovey commented on egs/wsj/s5/steps/dict/learn_lexicon.sh:
```diff
@@ -0,0 +1,402 @@
+  awk 'NR==FNR{a[$0] = 1; next} (!($0 in a))' $dir/pruned_prons.txt $dir/g2p_pron_candidates.txt \
+    > $dir/g2p_pron_candidates_pruned.txt \
+
+  cat $dir/phone_decoding_generated_prons_pruned.txt $dir/g2p_pron_candidates_pruned.txt \
```
for lines like this, there always needs to be a comment explaining what is going on.
also, the name 'dict_combined2' is not very informative.. I'm not sure what to suggest though.

xiaohui-zhang replied:
I changed it to "dict_combined_iter2" and the previous "dict_combined" to "dict_combined_iter1"

danpovey commented on egs/wsj/s5/steps/dict/learn_lexicon.sh:
```diff
@@ -0,0 +1,402 @@
+  else
+    echo "$0: The user requested not to apply the ref_lexicon_edits file. So we just exit at this point."
+    echo "  ... and leave it to the user to inspect/modify the edits file and then apply it on"
+    echo "  ... $dest_dict/lexicon0.txt to produce the final learned lexicon $dest_dict/lexicon.txt."
```
Please print out the exact command line the user must run.
Also, what's wrong with always applying the edits, but also always printing a message saying that you can modify the edits and then reapply them with such-and-such a command line?

xiaohui-zhang replied:
I've done as you suggested.

danpovey commented on egs/wsj/s5/steps/dict/prune_pron_candidates.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
```
For the scripts, like this one, where you think it's unlikely that a user will want to call them directly, please put them in a subdirectory internal/.  This makes it easier to look through the list of top-level scripts.

danpovey commented on egs/wsj/s5/steps/dict/prune_pron_candidates.py:
```diff
@@ -0,0 +1,120 @@
+                        "each word, based on which we prune the pron candidates.")
+    parser.add_argument("pruned_prons", metavar = "<pruned-prons>", type = str,
+                        help = "A file in lexicon format, which contains prons we want to" 
+                        "prune away from the pron_stats file.")
```
I think what you mean here is the prons we want to retain from the pron_stats file.
When you prune a tree, the parts that are pruned are the bits you cut off from the tree.

xiaohui-zhang replied:
hmm I meant pruned prons, not retained. It's more convenient for following stages if we output here the prons we want to throw away.

danpovey commented on egs/wsj/s5/steps/dict/select_prons_bayesian.py:
```diff
@@ -0,0 +1,415 @@
+            if pron not in learned_lexicon[word]:
+                print('R', ' | N | ', "%2.1f | " % stats.get((word, pron), 0), pron, 
+                      file=args.ref_lexicon_edits_handle)
+    print("Here are the words whose reference pron candidates were all declined", words[0]['100'])
```
Try to avoid extremely long lines in the source code.
Also, this would be much more readable if you set some variables with informative names instead of using those expressions directly, and used .format() instead of string concatenation.

danpovey commented:
Usage messages always need to make it easy to tell whether something is an input or an output parameter (e.g. does the script  write to, or read, the file?)

xiaohui-zhang commented:
Sure. I noted output file as "Output file containing XXX".

On Wed, Nov 9, 2016 at 5:49 PM, Daniel Povey <notifications@github.com>
wrote:

> Usage messages always need to make it easy to tell whether something is an
> input or an output parameter (e.g. does the script write to, or read, the
> file?)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-259549486>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANiEEQ5B8OGg3FVkPclFZozhlDrU9-LFks5q8k4LgaJpZM4JmzJX>
> .
>



-- 
Xiao-hui

xiaohui-zhang commented on egs/wsj/s5/steps/cleanup/internal/get_pron_stats.py:
```diff
@@ -0,0 +1,203 @@
+        # extract the nonsil_left and nonsil_right segments, and then try to
+        # append nonsil_left to the pron candidates of preceding word, getting
+        # extended pron candidates.
+        if word == '<eps>' or (word in non_scored_words and word != '<unk>' and word != '<UNK>'):
```
On Aug 27 I wrote in email:
Hi Dan, I noticed that if we allow all non_speech entries in the uncleaned r1 data, in the ctm_pron file, there are many cases like:

KevinStone_2010U-0024782-0025580-1 [UH] EH
KevinStone_2010U-0024782-0025580-1 fda F T
KevinStone_2010U-0024782-0025580-1 [NOISE] IY EY

So I'm thinking I should statistically merge the phones aligned to non_speech words to nearby words, as I did for <eps> before, since they behave like <eps>. Do you agree?

and you replied that makes sense. So I was trying to treat non-scored words (except OOV symbol <unk>) as <eps>. I'll put this example in the script.

xiaohui-zhang commented:
Thanks for @danpovey 's detailed suggestions. I've addressed them in my last commit. Please review when you have time :-)

xiaohui-zhang commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,134 @@
+data=data/train
+# acoustic training data we use to get alternative"
+# pronunciations and collet acoustic evidence."
+min_prob=0.4
```
It's the threshold we apply after normalizing the lexicon s.t. the max-pron-prob for each word is 1.0. I've clarified that.

danpovey commented on egs/wsj/s5/steps/dict/learn_lexicon.sh:
```diff
@@ -0,0 +1,405 @@
+if [ $stage -le 3 ]; then
+  echo "$0: Generate pronunciation candidates from phone-level decoding on acoustic training data.."
+  mdl=$src_mdl && $retrain_src_mdl && mdl=${src_mdl}_${affix}
+  steps/cleanup/debug_lexicon.sh --nj $nj --cmd "$decode_cmd" $data $dir/lang_expanded_train \
```
Something that I think you should investigate at some point (not necessarily right now), is whether you can improve this recipe by using a higher-order phone language model in steps/cleanup/debug_lexicon.sh.  Currently it uses an un-smoothed bigram (adding unigrams causes a blowup), but you could train a triphone LM using make_phone_lm.py using the same inputs (alignments); that also does not smooth below bigram.  You'd have to make sure the disambiguation symbol was taken care of... make_phone_lm.py uses a disambiguation symbol.  You could probably use `#0`, just make sure it's in phones.txt and L_disambig has the self-loop to consume it.

xiaohui-zhang replied:
sure I've added to my TODO list

danpovey commented on egs/wsj/s5/steps/dict/prons_to_lexicon.py:
```diff
@@ -0,0 +1,191 @@
+        else:
+            raise Exception("Unknown value {0} for --{1}".format(values, self.dest))
+
+def GetArgs():
```
We are moving the python to a pep8-compliant style (e.g. lower_case function names), and validation with flake8, but you could do that later.

xiaohui-zhang replied:
ok thanks

danpovey commented on egs/wsj/s5/steps/dict/prons_to_lexicon.py:
```diff
@@ -0,0 +1,191 @@
+        words.add(entry[0])
+        print("{0} {1}".format(entry[0], entry[1]),
+                file = args.out_lexicon_handle)
+    print ("Before pruning, the total num. pronunciations is: {}".format(len(lexicon)))
```
I think it's best to print things like this to stderr, as a general rule.  That way you can redirect the output to /dev/stdout and things will work.

danpovey commented on egs/wsj/s5/steps/dict/select_prons_bayesian.py:
```diff
@@ -0,0 +1,447 @@
+                    break
+            if not all_ref_prons_accepted or flags[0] == '1' or flags[2] == '1':
+                words_to_edit.append((word, counts[word]))
+            if count > thr:
```
please use threshold, not thr-- google style guide is against abbreviated names.

xiaohui-zhang commented on egs/wsj/s5/steps/dict/learn_lexicon.sh:
```diff
@@ -0,0 +1,405 @@
+src_mdl=$4
+ref_lang=$5
+dest_dict=$6
+dir=${src_mdl}_${affix}_work # Most outputs will be put here. 
```
ok so I'll just let the user to the set dir as something like ${src_mdl}_${affix} outside this script. is it fine?

xiaohui-zhang replied:
actually if we set dir outside the script, I think we can just remove the variable $affix (default to "lex", which now appears only in the dir name where put the re-trained model), and we can just put the re-trained model under $dir/${src_mdl}_retrained. Is it good?

danpovey replied:
That all sounds good [RE affix.]

xiaohui-zhang commented on egs/wsj/s5/steps/cleanup/internal/get_pron_stats.py:
```diff
@@ -0,0 +1,223 @@
+    parser.add_argument("silence_file", metavar = "<silphone-file>", type = str,
+                        help = "File containing a list of silence phones.")
+    parser.add_argument("optional_silence_file", metavar = "<optional_silence>", type = str,
+                        help = "File containing the optional silence phone. We'll be replacing empty prons by this.")
```
(from our old thread): 
> Sorry I forgot to mention, this is the candidate generating phase so we have
> to remove empty prons or merge them with SIL. So do you think we should
> merge them at this point? Actually I think simply removing empty prons may
> make sense also, though the file counts.txt generated later won't reflect
> the real counts.

Oh.  Yes, empty prons would cause a problem for lattice word
alignment.  So maybe turn them into SIL as you suggested.  I think we
should keep all the counts.

xiaohui-zhang replied:
The overall script doesn't output any pron containing silence phones. I printed out those prons in the log file when running select_prons_bayesian.py:
  print('WARNING: removing the candidate pronunciation from phonetic-decoding: {0}:'
    '{1} whose soft-count from lattice-alignment is {2} cause it contains at'
   'least one silence phone.'.format(word, pron, count))

jtrmal commented:
Guys, I won't have probably time for this until half of the december (until
LORELEI dry0run finishes)
I tried the scripts on babel data (Pastho) a couple of weeks back and it
actually crashed. Unfortunately I don't have the old system anymore and I
don't know what was the error.
I was wondering if @xiaohui-zhang could try to run the scripts and see if
he gets the error and if yes, then what is the reason for that. It might be
just that I was using older scripts and the bug was fixed since that. I'm
not saying he has to make actually improving something -- just to see if
there is some robustness issue
Thanks,
y.

On Wed, Aug 17, 2016 at 3:26 PM, Xiaohui Zhang <notifications@github.com>
wrote:

> the main script is egs/wsj/s5/steps/dict/learn_lexicon.sh
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kaldi-asr/kaldi/pull/976
> Commit Summary
>
>    - WIP: lexicon learning
>
> File Changes
>
>    - *A* egs/tedlium/s5_r2/local/run_learn_lex.sh
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-0> (79)
>    - *M* egs/wsj/s5/steps/cleanup/debug_lexicon.sh
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-1> (30)
>    - *A* egs/wsj/s5/steps/cleanup/internal/get_pron_stats.py
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-2> (174)
>    - *A* egs/wsj/s5/steps/dict/apply_g2p.sh
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-3> (127)
>    - *A* egs/wsj/s5/steps/dict/get_pron_stats.py
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-4> (91)
>    - *A* egs/wsj/s5/steps/dict/learn_lexicon.sh
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-5> (341)
>    - *A* egs/wsj/s5/steps/dict/prons_to_lexicon.py
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-6> (193)
>    - *A* egs/wsj/s5/steps/dict/select_prons_bayesian.py
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-7> (354)
>    - *A* egs/wsj/s5/steps/dict/train_g2p.sh
>    <https://github.com/kaldi-asr/kaldi/pull/976/files#diff-8> (94)
>
> Patch Links:
>
>    - https://github.com/kaldi-asr/kaldi/pull/976.patch
>    - https://github.com/kaldi-asr/kaldi/pull/976.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX8vu6v1K9KKoAZYkyx68erk64rQ7ks5qg2B0gaJpZM4JmzJX>
> .
>

xiaohui-zhang commented on egs/wsj/s5/steps/dict/internal/prune_pron_candidates.py:
```diff
@@ -0,0 +1,120 @@
+import math
+
+def GetArgs():
+    parser = argparse.ArgumentParser(description = "Prune pronunciation candidates based on soft-counts from lattice-alignment"
```
no. I made sure in the script that once one pron appears in the reference lexicon, I never prune it away.

xiaohui-zhang commented:
@jtrmal Thanks. I'll make sure this pipeline works. After you finish the LORELEI dryrun the priority of getting your help might be experiments related to the new g2p toolkit. 

@danpovey just addressed all your comments (and tested the whole thing)

danpovey commented:
So you think this is ready to merge?

On Wed, Nov 16, 2016 at 5:14 PM, Xiaohui Zhang <notifications@github.com>
wrote:

> @jtrmal <https://github.com/jtrmal> Thanks. I'll make sure this pipeline
> works. After you finish the LORELEI dryrun the priority of getting your
> help might be experiments related to the new g2p toolkit.
>
> @danpovey <https://github.com/danpovey> just addressed all your comments
> (and tested the whole thing)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-261089551>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu1zCf1hpLtELyfkXy6Q2BBZHMAPIks5q-4BFgaJpZM4JmzJX>
> .
>

xiaohui-zhang commented:
yep. as long as you don't have other suggestions

danpovey commented on egs/tedlium/s5_r2/local/run_learn_lex.sh:
```diff
@@ -0,0 +1,135 @@
+       data/${dset} exp/tri3_learned_lex/decode_${dset} exp/tri3_learned_lex/decode_${dset}_rescore || exit 1;
+  ) &
+  done
+fi
```
Can you put results somewhere in the script, so I can see what the WER is before and after?

vince62s commented:
@xiaohui-zhang Do you have results to commit too ?

danpovey commented:
I'm asking for some big changes before we commit this, but Samuel, you
might as well put the results on this PR.
Basically, after looking at the lexicon-edits it was making, I think the
algorithm should be changed to use a different objective function that's
more related to likelihood...


On Thu, Nov 17, 2016 at 1:57 AM, vince62s <notifications@github.com> wrote:

> @xiaohui-zhang <https://github.com/xiaohui-zhang> Do you have results to
> commit too ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/976#issuecomment-261170616>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu2EDaGJPFHUUPFauOeXMbIv7m9YBks5q-_rqgaJpZM4JmzJX>
> .
>

danpovey commented:
Since it looks like the alternative version we're looking into may be becoming a bigger and more longer-term project, I'm considering committing this as a stopgap measure.  Samuel, are there any small fixes that you need to apply to this PR?

xiaohui-zhang commented:
@danpovey good to merge now

xiaohui-zhang commented:
Hello @vince62s , sorry for the late reply. We've been working hard on improving WERs and it's hard to get consistent improvements, partly cause the pronunciations we improved don't always appear in the dev/test sets. Also my baseline was out-of-date. So I didn't bother putting numbers in RESULTS. I'll let you know when we have consistent WER improvements. Thanks!

vince62s commented:
@xiaohui-zhang I didn't try the scripts nor read it carefully because I am not sure to fully understand what it does in the case of tedlium. I would appreciate some clarification.

In the case of Tedlium, unless I am wrong, training data do not have OOV, since they included all words from training data in the dictionary. so, if this is the case, am I right saying that this script step will do nothing ?

from another general standpoint, if a user does the job on his side (to use G2P to generate prons for oov words in the training data) which I guess is what the tedlium team did when built their dict, will the script do some kind of redundant thing / detect some lexicon errors that were initiated by G2P in the first place ?

thanks for your help.

xiaohui-zhang commented:
Hi @vince62s , basically there is OOV cause the vocab size of data/local/dict/lexicon.txt was truncated to 200k. This recipe is not showing how to deal with OOVs with lexicon learning. What this recipe does is: it debugs the given reference lexicon (data/local/dict/lexicon.txt), so that the pronunciations agree more with the acoustics. The output lexicon will be put in data/local/dict_learned_nosp (Note that the vocab is the same as the reference lexicon, cause "target-vocab" in learn_lexicon.sh was set as ref-vocab (vocab of the reference lexicon. The recipe also outputs a lexicon-edit file, which records the changes to the reference lexicon).
